classes/ActionScheduler_wpPostStore.php

```diff
@@ -462,25 +465,93 @@ protected function generate_claim_id() {
+	 *
 	 * @return int The number of actions that were claimed
 	 * @throws RuntimeException
 	 */
-	protected function claim_actions( $claim_id, $limit, DateTime $before_date = NULL ) {
+	protected function claim_actions( $claim_id, $limit, DateTime $before_date = null, $group = '' ) {
 		/** @var wpdb $wpdb */
 		global $wpdb;
 
-		$date = is_null($before_date) ? as_get_datetime_object() : clone $before_date;
+		$date = null === $before_date ? as_get_datetime_object() : clone $before_date;
+
+		$query      = new WP_Query();
+		$query_args = array(
+			'fields'         => 'ids',
+			'post_type'      => self::POST_TYPE,
+			'post_status'    => ActionScheduler_Store::STATUS_PENDING,
+			'has_password'   => false,
+			'posts_per_page' => $limit,
+			'orderby'        => array(
+				'menu_order' => 'ASC',
+				'date'       => 'ASC',
+				'ID'         => 'ASC',
+			),
+			'date_query'     => array(
+				'column' => 'post_date_gmt',
+				array(
+					'compare' => '<=',
+					'year'    => $date->format( 'Y' ),
+					'month'   => $date->format( 'n' ),
+					'day'     => $date->format( 'j' ),
+					'hour'    => $date->format( 'G' ),
+					'minute'  => $date->format( 'i' ),
+					'second'  => $date->format( 's' ),
+				)
+			),
+		);
+
+		// Check for a particular group.
+		if ( ! empty( $group ) && term_exists( $group, self::GROUP_TAXONOMY ) ) {
+			$query_args['tax_query'] = array(
+				array(
+					'taxonomy'         => self::GROUP_TAXONOMY,
+					'field'            => 'slug',
+					'terms'            => $group,
+					'include_children' => false,
+				),
+			);
+		}
+
+		// Gather relevant IDs
+		$ids = $query->query( $query_args );
+
 		// can't use $wpdb->update() because of the <= condition, using post_modified to take advantage of indexes
-		$sql = "UPDATE {$wpdb->posts} SET post_password = %s, post_modified_gmt = %s, post_modified = %s WHERE post_type = %s AND post_status = %s AND post_password = '' AND post_date_gmt <= %s ORDER BY menu_order ASC, post_date_gmt ASC, ID ASC LIMIT %d";
-		$sql = $wpdb->prepare( $sql, array( $claim_id, current_time('mysql', true), current_time('mysql'), self::POST_TYPE, 'pending', $date->format('Y-m-d H:i:s'), $limit ) );
-		$rows_affected = $wpdb->query($sql);
+		$update = "UPDATE {$wpdb->posts} SET post_password = %s, post_modified_gmt = %s, post_modified = %s";
+		$params = array(
+			$claim_id,
+			current_time( 'mysql', true ),
+			current_time( 'mysql' ),
+		);
+
+		// Build initial WHERE clause
+		$where  = "WHERE post_type = %s AND post_status = %s AND post_password = ''";
+		$params[] = self::POST_TYPE;
+		$params[] = ActionScheduler_Store::STATUS_PENDING;
+
+		// If we have IDs, include them in the WHERE clause. Otherwise use the old date parameter.
+		if ( count( $ids ) > 0 ) {
+			$where .= ' AND ID IN (' . join( ',', $ids ) . ')';
+//		} else {
+//			$where    .= ' AND post_date_gmt <= %s';
+//			$params[] = $date->format( 'Y-m-d H:i:s' );
```

Comment:
@thenbrent I'd like to solicit some feedback on what I've been doing here. Here's my thought process on how this should be updated to work:

1. To enable querying via a group (which in this case is a taxonomy), I though it would make the most sense to use `WP_Query` if possible. In this case, we would be querying for IDs using the taxonomy.
1. With a set of IDs, we could then limit our own query to the set of IDs we identified, and thus we accomplish the limit by group.

Here's where I'm having a problem, and I think that I'm stuck enough to question whether I'm even approaching this the right way.

Using `WP_Query` requires that we recreate the same conditions that are being applied using `$wpdb->query()` below. Otherwise, we may end up with post IDs that don't match. This leads me to setting `post_type`, `post_status`, and also `date_query`. The problem is in using `date_query`, it's not quite exactly what we use in our condition below. *In theory* it should end up with the same results, but of course that's not a good assumption.

In trying to avoid making assumptions, I've been actively running our existing Unit Tests to ensure I'm not breaking anything, and that has been very helpful. Where I'm at now is that there is a particular test that keeps failing: `ActionScheduler_QueueRunner_Test::test_next_instance_of_action()`. Well, it fails so long as these three lines remain commented out.

This tells me that the my `WP_Query` parameters are ending up with an empty result, which leads me to think that I'm doing something incorrectly. Or, I'm otherwise missing something.

I'm open to any thoughts you have, whether it's a different approach entirely or a better way to run `WP_Query`. I've also tried something like this for `date_query`, with even less successful results:

```
			'date_query'     => array(
				array(
					'column' => 'post_date_gmt',
					'compare' => '<=',
					'before' => $date->format( 'Y-m-d H:i:s' ),
					'inclusive' => true,
				),
			),
```

Reply:
Hrm, both of those date queries look OK to me. At least based on revising the [WP Query](https://codex.wordpress.org/Class_Reference/WP_Query#Date_Parameters) and [`WP_Date_Query`](https://developer.wordpress.org/reference/classes/wp_date_query/) docs.

I double checked that `WP_Date_Query` supported `<=` because of the comment in the existing code about not using `$wpdb->update()` because it doesn't support `<=`, but it looks fine too.

The `'tax_query'` also looks good.

It's unlikely, but something else could be adding another `tax_query` or `date_query` (via a lower level WP filter), and because the default `relation` is `AND`, it's returning empty results. That's so unlikely it feels barely worth mentioning, but I'm not seeing any more likely answers for the reason it's not working. Other than some obvious error we're both missing around syntax of the `tax_query` or `date_query` (I often get tripped up on the multidimensional arrays).

I think this is a good approach. If you want to explore an alternative seeing as you can't get it to work yet, you could try using a subquery in the existing `UPDATE` query. That would sure be easier to test than using `WP_Query` as you can just run the query directly as SQL. It's not as simple as it could be though given the disjointed schema for taxonomies and terms.